File: app/src/main/java/org/pragyan/dalal18/notifications/NotificationFragment.java
package org.pragyan.dalal18.notifications;


import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.pragyan.dalal18.R;
import org.pragyan.dalal18.adapter.NotificationRecyclerAdapter;
import org.pragyan.dalal18.dagger.ContextModule;
import org.pragyan.dalal18.dagger.DaggerDalalStreetApplicationComponent;
import org.pragyan.dalal18.data.Notification;
import org.pragyan.dalal18.loaders.NotificationLoader;
import org.pragyan.dalal18.utils.ConnectionUtils;
import org.pragyan.dalal18.utils.Constants;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import dalalstreet.api.DalalActionServiceGrpc;
import dalalstreet.api.actions.GetNotificationsResponse;

public class NotificationFragment extends Fragment implements LoaderManager.LoaderCallbacks<GetNotificationsResponse> {

    private static final String LAST_NOTIFICATION_ID = "last_notification_id";

    @Inject
    DalalActionServiceGrpc.DalalActionServiceBlockingStub actionServiceBlockingStub;

    @Inject
    NotificationRecyclerAdapter notificationRecyclerAdapter;

    @Inject
    SharedPreferences preferences;

    @BindView(R.id.notifications_recyclerView)
    RecyclerView notificationsRecyclerView;

    @BindView(R.id.noNotification_textView)
    TextView noNotificationTextView;

    private ConnectionUtils.OnNetworkDownHandler networkDownHandler;
    private AlertDialog loadingDialog;

    boolean paginate = true;
    ArrayList<Notification> customNotificationList = new ArrayList<>();

    public NotificationFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            networkDownHandler = (ConnectionUtils.OnNetworkDownHandler) context;
        } catch (ClassCastException classCastException) {
            throw new ClassCastException(context.toString() + " must implement network down handler.");
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getContext() != null) {
            View dialogView = LayoutInflater.from(getContext()).inflate(R.layout.progress_dialog, null);
            String tempString = "Getting notifications...";
            ((TextView) dialogView.findViewById(R.id.progressDialog_textView)).setText(tempString);
            loadingDialog = new AlertDialog.Builder(getContext())
                    .setView(dialogView)
                    .setCancelable(false)
                    .create();
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_notification, container, false);
        ButterKnife.bind(this, rootView);

        if (getActivity() != null) getActivity().setTitle(getString(R.string.notifications));
        DaggerDalalStreetApplicationComponent.builder().contextModule(new ContextModule(getContext())).build().inject(this);

        notificationRecyclerAdapter = new NotificationRecyclerAdapter(getContext(), null);
        notificationsRecyclerView.setAdapter(notificationRecyclerAdapter);
        notificationsRecyclerView.setHasFixedSize(false);
        notificationsRecyclerView.addOnScrollListener(new CustomScrollListener());
        notificationsRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        getActivity().getSupportLoaderManager().restartLoader(Constants.NOTIFICATION_LOADER_ID, null, this);
        loadingDialog.show();

        return rootView;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        preferences.edit().remove(LAST_NOTIFICATION_ID).apply();
    }

    @Override
    public Loader<GetNotificationsResponse> onCreateLoader(int id, Bundle args) {

        if (getContext() != null) {
            int lastId = preferences.getInt(LAST_NOTIFICATION_ID, 0);
            return new NotificationLoader(getContext(), actionServiceBlockingStub, lastId);
        } else {
            return null;
        }
    }

    @Override
    public void onLoadFinished(Loader<GetNotificationsResponse> loader, GetNotificationsResponse data) {

        loadingDialog.dismiss();

        if (data == null) {
            networkDownHandler.onNetworkDownError();
            return;
        }

        paginate = data.getNotificationsCount() == 10;

        if (data.getNotificationsList().size() > 0) {

            for (dalalstreet.api.models.Notification currentNotification : data.getNotificationsList()) {
                customNotificationList.add(new Notification(currentNotification.getText(), currentNotification.getCreatedAt()));
                preferences.edit()
                        .putInt(LAST_NOTIFICATION_ID, currentNotification.getId())
                        .apply();
            }

            notificationRecyclerAdapter.swapData(customNotificationList);
            noNotificationTextView.setVisibility(View.GONE);
            notificationsRecyclerView.setVisibility(View.VISIBLE);

        } else {
            noNotificationTextView.setVisibility(View.VISIBLE);
            notificationsRecyclerView.setVisibility(View.GONE);
        }
    }

    @Override
    public void onLoaderReset(Loader<GetNotificationsResponse> loader) {
        // Do nothing
    }


    public class CustomScrollListener extends RecyclerView.OnScrollListener {

        CustomScrollListener() {
        }

        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            int visibleItemCount = recyclerView.getLayoutManager().getChildCount();
            int totalItemCount = recyclerView.getLayoutManager().getItemCount();
            int pastVisibleItems = ((LinearLayoutManager) recyclerView.getLayoutManager()).findFirstVisibleItemPosition();
            if (pastVisibleItems + visibleItemCount >= totalItemCount) {

                if (paginate) {
                    if (getActivity() != null) {
                        getActivity().getSupportLoaderManager().restartLoader(Constants.NOTIFICATION_LOADER_ID, null, NotificationFragment.this);
                        paginate = false;
                    }
                }
            }
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        preferences.edit().remove(LAST_NOTIFICATION_ID).apply();
    }
}

File: app/src/main/java/org/pragyan/dalal18/notifications/NotificationFragment.kt
package org.pragyan.dalal18.notifications

import android.content.*
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.LinearLayoutManager
import dalalstreet.api.DalalActionServiceGrpc
import dalalstreet.api.actions.GetNotificationsRequest
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import org.pragyan.dalal18.R
import org.pragyan.dalal18.adapter.NotificationRecyclerAdapter
import org.pragyan.dalal18.dagger.ContextModule
import org.pragyan.dalal18.dagger.DaggerDalalStreetApplicationComponent
import org.pragyan.dalal18.data.Notification
import org.pragyan.dalal18.databinding.FragmentNotificationBinding
import org.pragyan.dalal18.ui.MainActivity
import org.pragyan.dalal18.utils.ConnectionUtils
import org.pragyan.dalal18.utils.Constants
import org.pragyan.dalal18.utils.viewLifecycle
import javax.inject.Inject

class NotificationFragment : Fragment() {

    private var binding by viewLifecycle<FragmentNotificationBinding>()

    @Inject
    lateinit var actionServiceBlockingStub: DalalActionServiceGrpc.DalalActionServiceBlockingStub

    @Inject
    lateinit var notificationRecyclerAdapter: NotificationRecyclerAdapter

    @Inject
    lateinit var preferences: SharedPreferences

    lateinit var networkDownHandler: ConnectionUtils.OnNetworkDownHandler
    private lateinit var loadingDialog: AlertDialog

    private var customNotificationList = ArrayList<Notification>()

    // Pagination
    private var firstFetch = true
    private var paginate = true
    private var lastId = 0

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            networkDownHandler = context as ConnectionUtils.OnNetworkDownHandler
        } catch (classCastException: ClassCastException) {
            throw ClassCastException("$context must implement network down handler.")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentNotificationBinding.inflate(inflater, container, false)
        DaggerDalalStreetApplicationComponent.builder().contextModule(ContextModule(context!!)).build().inject(this)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val dialogView = LayoutInflater.from(context).inflate(R.layout.progress_dialog, null)
        val tempString = "Getting notifications..."
        (dialogView.findViewById<View>(R.id.progressDialog_textView) as TextView).text = tempString
        loadingDialog = AlertDialog.Builder(context!!)
                .setView(dialogView)
                .setCancelable(false)
                .create()

        buildRecyclerView()

        preferences.edit().remove(Constants.LAST_NOTIFICATION_ID).apply()
        getNotificationsAsynchronously()
    }

    private fun buildRecyclerView() {
        notificationRecyclerAdapter = NotificationRecyclerAdapter(context, null)
        binding.notificationsRecyclerView.apply {
            adapter = notificationRecyclerAdapter
            setHasFixedSize(false)
            layoutManager = LinearLayoutManager(this@NotificationFragment.context)
        }

        binding.mainContent.setOnScrollChangeListener(NestedScrollView.OnScrollChangeListener { v, _, scrollY, _, _ ->
            if (scrollY == v.getChildAt(0).measuredHeight - v.measuredHeight) {
                if (paginate && activity != null) {
                    println("Bottom")
                    getNotificationsAsynchronously()
                    paginate = false
                }
            }
        })
    }

    private fun getNotificationsAsynchronously() {
        if (firstFetch)
            loadingDialog.show()
        lastId = preferences.getInt(Constants.LAST_NOTIFICATION_ID, 0)

        doAsync {
            if (ConnectionUtils.getConnectionInfo(context)) {
                if (ConnectionUtils.isReachableByTcp(Constants.HOST, Constants.PORT)) {
                    val notificationsResponse = actionServiceBlockingStub.getNotifications(GetNotificationsRequest
                            .newBuilder()
                            .setLastNotificationId(lastId)
                            .setCount(10)
                            .build())

                    uiThread {
                        paginate = notificationsResponse.notificationsCount == 10
                        if (!paginate) {
                            // No more Transactions left to show for the user
                            binding.progressBar.visibility = View.GONE
                        }

                        val positionStart = customNotificationList.size
                        val newItemsCount = notificationsResponse.notificationsList.size

                        customNotificationList.addAll(notificationsResponse.notificationsList.map {
                            Notification(it.text, it.createdAt)
                        })

                        binding.apply {
                            if (firstFetch) {
                                if (customNotificationList.isEmpty()) {
                                    // No Notifications on the First Fetch itself
                                    noNotificationTextView.visibility = View.VISIBLE
                                } else
                                    mainContent.visibility = View.VISIBLE

                                firstFetch = false
                            }
                        }
                        notificationRecyclerAdapter.setList(customNotificationList)
                        notificationRecyclerAdapter.notifyItemRangeInserted(positionStart, newItemsCount)

                        preferences.edit()
                                .putInt(Constants.LAST_NOTIFICATION_ID, notificationsResponse.notificationsList.last().id)
                                .apply()
                    }
                } else {
                    uiThread { networkDownHandler.onNetworkDownError(resources.getString(R.string.error_server_down), R.id.notifications_dest) }
                }
            } else {
                uiThread { networkDownHandler.onNetworkDownError(resources.getString(R.string.error_check_internet), R.id.notifications_dest) }
            }
            if (firstFetch)
                uiThread { loadingDialog.dismiss() }
        }
    }

    private val refreshNotifications = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            customNotificationList.add(Notification(intent.getStringExtra(TEXT_KEY), intent.getStringExtra(CREATED_AT_KEY)))
            customNotificationList.sortByDescending { it.createdAt }
            notificationRecyclerAdapter.setList(customNotificationList)
            notificationRecyclerAdapter.notifyDataSetChanged()
        }
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(refreshNotifications)
        preferences.edit().remove(Constants.LAST_NOTIFICATION_ID).apply()
    }

    override fun onResume() {
        super.onResume()
        val intentFilter = IntentFilter(MainActivity.REFRESH_UNREAD_NOTIFICATIONS_COUNT)
        LocalBroadcastManager.getInstance(requireContext()).registerReceiver(refreshNotifications, intentFilter)
    }

    override fun onDestroy() {
        super.onDestroy()
        preferences.edit().remove(Constants.LAST_NOTIFICATION_ID).apply()
    }

    companion object {
        const val TEXT_KEY = "text-key"
        const val CREATED_AT_KEY = "createdat-key"
    }
}
